syft/pkg/cataloger/java/parse_pom_xml.go

```diff
@@ -196,16 +219,25 @@ func collectDependencies(ctx context.Context, r *maven.Resolver, resolved map[ma
 			}
 		}
 
-		pkgs = append(pkgs, *depPkg)
+		pkgs = append(pkgs, depPkg)
 		if parentPkg != nil {
-			relationships = append(relationships, artifact.Relationship{
-				From: *depPkg,
-				To:   *parentPkg,
+			relationships = append(relationships, artifact.Relationship{ //nolint:gocritic // we intentionally want to use the reference to the package which will still be mutated and finalized to a value later
+				// both the to and from packages may be mutated based on the resolved dependencies, so we need references to these values
+				// to ensure the nodes used are consistent with the final state of the packages.
+				// note: it is VITAL that these references are replaced with values by the caller of this function before using
```

Comment:
These kind of comments always make me nervous. Is there something we can do to make sure this happens? What if we return something that represents instructions for building an `[]artifact.Relationships` in the caller, so that the caller can't forget and return a `[]artifact.Relationships` with unexpected types in it?

Reply:
yes that's what the `finalizeRelationships` call at the final return of the cataloger (a chokepoint) https://github.com/anchore/syft/blob/5b7ec60f8d13f7a211aa4a799bcbd84edd1c35af/syft/pkg/cataloger/java/parse_pom_xml.go#L84-L97

Reply:
I see that. I think I'd rather the type system remember that `finalizeRelationships` needs to be called. For example, you could have a local struct `unfinalizedRelationship` and return a slice of those, and then finalize relationship can take a slice of `unfinalizedRelationship` and return a slice of a `artifact.Relationship`. That way if someone tries to use this function without finalizing the relationships, the compiler won't let them.